lib/smack/SmackInstGenerator.cpp

```diff
@@ -144,6 +144,33 @@ void SmackInstGenerator::processInstruction(llvm::Instruction& inst) {
   nextInst++;
 }
 
+void SmackInstGenerator::processIntrinsicCall(llvm::IntrinsicInst* ii) {
+  static auto f16Cast = [&ii, this](bool f16UpCast) {
```

Comment:
Oh, man, this is maybe overly complicated just to avoid using the names of the intrinsics. Sorry about that.
Let's actually use their names in combination with `getIntrinsicID`. So you first switch based on `getIntrinsicID` to get into the right top-level intrinsic type, and then you switch based on names. The list of names will be exhaustive for a particular intrinsic ID, and there will be an assertion in the code checking that a name gets matches. This assertion will catch any changes in the names of the intrinsics.

Reply:
Despite (or because of) how hard this was to get type-checked... I kind of agree.

Reply:
I actually prefer not to use function names now as they might not be a reliable and are subject to version changes. Using function names is essentially more or less equivalent to what we're doing now, which is finding the precise intrinsic associated with the overloaded ID by using function signatures.

Reply:
@zvonimir do you mean the complexity of using the lambda map compared to that of using a switch statement?
I think the IDs in the enum are very unlikely to be changed.